libyul/ObjectOptimizer.h

```diff
@@ -0,0 +1,97 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/ASTForward.h>
+#include <libyul/Object.h>
+
+#include <liblangutil/EVMVersion.h>
+
+#include <libsolutil/FixedHash.h>
+
+#include <map>
+#include <memory>
+#include <optional>
+
+namespace solidity::yul
+{
+
+enum class Language
+{
+	Yul,
+	Assembly,
+	StrictAssembly,
+};
+
+Dialect const& languageToDialect(Language _language, langutil::EVMVersion _version);
+
+/// Encapsulates logic for applying @a yul::OptimiserSuite to a whole hierarchy of Yul objects.
+/// Also, acts as a transparent cache for optimized objects.
+///
+/// The cache is designed to allow sharing its instances widely across the compiler, without the
+/// need to invalidate entries due to changing settings or context.
+/// Caching is performed at the granularity of individual ASTs rather than whole object trees,
+/// which means that reuse is possible even within a single hierarchy, e.g. when creation and
+/// deployed objects have common dependencies.
+class ObjectOptimizer
+{
+public:
+	/// Optimization settings and context information.
+	/// This information becomes a part of the cache key and, together with the object content,
+	/// must uniquely determine the result of optimization.
+	struct Settings
+	{
+		Language language;
+		langutil::EVMVersion evmVersion;
+		bool optimizeStackAllocation;
+		std::string yulOptimiserSteps;
+		std::string yulOptimiserCleanupSteps;
```

Comment:
It's just all over the place. It's a losing battle :) I used to try to be consistent and use BE spelling for this but in the end gave up because we already use both in different places.

And also we decided some time ago to go with `-ize` because BE allows that too so that's what I'm using nowadays when naming new things.

In this particular case these values are already named with `-ise` in `OptimizerSettings` so I'm just sticking to that to at least keep it consistent between these structs.

Reply:
```libyul/optimiser/OptimizerUtilities.h`` containing ``Small useful snippets for the optimiser.`` - at least it's alternating :-).